### PR TITLE
server: Read PKSC8 private keys

### PR DESCRIPTION
**Pull request by olix0r**
Currently the server only attempts to parse RSA private keys. This
change updates the server to first attempt to read PKSC8 private keys
before attempting to parse the key file as RSA.

### CODE REVIEW

File: kubert/src/server.rs [hunk 1]
```diff
@@ -298,7 +300,7 @@ impl TlsCertPath {
     async fn load_certs(&self) -> std::io::Result<Vec<rustls::Certificate>> {
         // Open certificate file.
         let pem = tokio::fs::read(&self.0).await?;
-        let mut reader = std::io::BufReader::new(pem.as_slice());
+        let mut reader = BufReader::new(pem.as_slice());
```
Comment (hawkw) [inline kubert/src/server.rs:303]: `BufRead` is [implemented for `&[u8]`](https://doc.rust-lang.org/nightly/std/io/trait.BufRead.html#impl-BufRead), so i don't _think_ the `BufReader` is necessary --- the whole thing is already buffered?
```suggestion
        let mut reader = pem.as_slice();
```

File: kubert/src/server.rs [hunk 2]
```diff
@@ -320,10 +322,10 @@ impl TlsKeyPath {
     async fn load_private_key(&self) -> std::io::Result<rustls::PrivateKey> {
         // Open keyfile.
         let pem = tokio::fs::read(&self.0).await?;
-        let mut reader = std::io::BufReader::new(pem.as_slice());
 
         // Load and return a single private key.
-        let keys = rustls_pemfile::rsa_private_keys(&mut reader)?;
+        let keys = rustls_pemfile::pkcs8_private_keys(&mut BufReader::new(pem.as_slice()))
+            .or_else(|_| rustls_pemfile::rsa_private_keys(&mut BufReader::new(pem.as_slice())))?;
```
Comment (hawkw) [inline kubert/src/server.rs:328]: similarly:
```suggestion
        let keys = rustls_pemfile::pkcs8_private_keys(&mut pem.as_slice())
            .or_else(|_| rustls_pemfile::rsa_private_keys(&mut pem.as_slice()))?;
```

File: kubert/src/server.rs [hunk 3]
```diff
@@ -4,7 +4,9 @@
 //! Unlike a normal `hyper` server, this server reloads its TLS credentials for each connection to
 //! support certificate rotation.
 
-use std::{convert::Infallible, net::SocketAddr, path::PathBuf, str::FromStr, sync::Arc};
+use std::{
+    convert::Infallible, io::BufReader, net::SocketAddr, path::PathBuf, str::FromStr, sync::Arc,
```
Comment (hawkw) [inline kubert/src/server.rs:8]: if we follow my other suggestions, i don't think we need the `BufReader` import?
```suggestion
    convert::Infallible, net::SocketAddr, path::PathBuf, str::FromStr, sync::Arc,
```

File: kubert/src/server.rs (hunk 2 above)
Comment (hawkw) [inline kubert/src/server.rs:328]: do we want to maybe log something about which format the keys were parsed in?

File: kubert/src/server.rs [hunk 4]
```diff
@@ -320,10 +322,10 @@ impl TlsKeyPath {
     async fn load_private_key(&self) -> std::io::Result<rustls::PrivateKey> {
         // Open keyfile.
         let pem = tokio::fs::read(&self.0).await?;
-        let mut reader = std::io::BufReader::new(pem.as_slice());
 
         // Load and return a single private key.
-        let keys = rustls_pemfile::rsa_private_keys(&mut reader)?;
+        let keys = rustls_pemfile::pkcs8_private_keys(&mut BufReader::new(pem.as_slice()))
```
Comment (adleong) [inline kubert/src/server.rs:327]: Looking at the implementation, I think these functions will return an empty vec if the file only contains the opposite key type, not an error. 

https://docs.rs/rustls-pemfile/latest/src/rustls_pemfile/lib.rs.html#75-85

If the vec of pkcs8 keys is empty, we want to look for rsa keys.

**Review by hawkw — APPROVED** [review]
looks good to me --- one suggestion.

File: kubert/src/server.rs (hunk 4 above)
Comment (olix0r) [reply kubert/src/server.rs]: Ah yeah, great catch!

File: kubert/src/server.rs (hunk 2 above)
Comment (olix0r) [reply kubert/src/server.rs]: I could go either way. for now I'll trust that whatever rustls logs is good enough

**Review by hawkw — APPROVED** [review]

**Pull request merged**